### PR TITLE
Add cli and API DID deactivation

### PR DESCRIPTION
**Pull request by stevenvegt**
Adds the CLI and API `Deactivate` commands.
Adds the deactivated status to the resolvemetadata.

### CODE REVIEW

File: docs/_static/vdr/v1.yaml [hunk 1]
```diff
@@ -100,7 +100,32 @@ paths:
             text/plain:
               schema:
                 type: string
-
+    delete:
+      summary: Deactivatates a Nuts DID Document according to the specification.
```
Comment (reinkrul) [inline docs/_static/vdr/v1.yaml:104]: I'd almost think you're yoking about my speech disorder.... :-p

File: docs/_static/vdr/v1.yaml [hunk 2]
```diff
@@ -100,7 +100,32 @@ paths:
...
+      operationId: "deactivateDID"
+      tags:
+        - DID
+      responses:
+        "200":
+          description: DID Document has been deactivated.
```
Comment (reinkrul) [inline docs/_static/vdr/v1.yaml:110]: DID or DID Document?

When there's no content returned, consider returning HTTP `201` (No Content)

File: docs/_static/vdr/v1.yaml [hunk 3]
```diff
@@ -100,7 +100,32 @@ paths:
...
+        "400":
+          description: DID couldn't be deactivated because the input conflicts or is malformed.
```
Comment (reinkrul) [inline docs/_static/vdr/v1.yaml:112]: conflict = HTTP 409

File: docs/_static/vdr/v1.yaml [hunk 4]
```diff
@@ -100,7 +100,32 @@ paths:
...
+          content:
+            text/plain:
+              schema:
+                type: string
+        "404":
+          description: Corresponding DID document could not be found.
```
Comment (reinkrul) [inline docs/_static/vdr/v1.yaml:118]: can we decide on DID Document or DID document?

File: vdr/cmd/cmd.go [hunk 5]
```diff
@@ -153,6 +155,24 @@ func resolveCmd() *cobra.Command {
 	return result
 }
 
+func deactivateCmd() *cobra.Command {
+	result := &cobra.Command{
+		Use:   "deactivate [DID]",
+		Short: "Deactivate a DID document based on its DID",
```
Comment (reinkrul) [inline vdr/cmd/cmd.go:161]: maybe mention this is a destructive operation and it can't be reverted

File: vdr/cmd/cmd_test.go [hunk 6]
```diff
@@ -240,6 +240,47 @@ func TestEngine_Command(t *testing.T) {
...
 		})
 	})
+
+	t.Run("deactivate", func(t *testing.T) {
+		t.Run("ok", func(t *testing.T) {
+			cmd := Cmd()
+			s := httptest.NewServer(http2.Handler{StatusCode: http.StatusOK})
+			os.Setenv("NUTS_ADDRESS", s.URL)
+			defer os.Unsetenv("NUTS_ADDRESS")
+			core.NewServerConfig().Load(cmd)
```
Comment (reinkrul) [inline vdr/cmd/cmd_test.go:250]: I think you should be using ClientConfig?

File: vdr/vdr.go [hunk 7]
```diff
@@ -170,11 +170,15 @@ func isDeactivated(document *did.Document) bool {
...
 // Deactivate updates the DID Document so it can no longer be updated
-func (r *VDR) Deactivate(id did.DID, currentHash hash.SHA256Hash) error {
+func (r *VDR) Deactivate(id did.DID) error {
+	_, meta, err := r.store.Resolve(id, &types.ResolveMetadata{AllowDeactivated: true})
+	if err != nil {
+		return err
+	}
 	// A deactivated document is the original document stripped from a controller and keys.
-	// So, apart from the services, it is practically an empty document. Should we even keep the services?
+	// So, apart from the services, it is practically an empty document.
```
Comment (reinkrul) [inline vdr/vdr.go:179]: It _is_ an empty document, even the services are removed (comment is incorrect)

**Review by reinkrul — APPROVED** [review]

**Review by reinkrul — CHANGES_REQUESTED** [review]
s

File: docs/_static/vdr/v1.yaml (hunk 3 above)
Comment (stevenvegt) [reply docs/_static/vdr/v1.yaml:112]: 409 is rare and I would not recommend using it.

File: docs/_static/vdr/v1.yaml (hunk 2 above)
Comment (stevenvegt) [reply docs/_static/vdr/v1.yaml:110]: 201 is created. Perhaps you mean 204? I would like to keep the set of status codes small. 200 seems a legit reply code.

File: vdr/cmd/cmd.go (hunk 5 above)
Comment (stevenvegt) [reply vdr/cmd/cmd.go:162]: Added a confirmation 👍

File: docs/_static/vdr/v1.yaml (hunk 3 above)
Comment (reinkrul) [reply docs/_static/vdr/v1.yaml]: I disagree but let's agree to disagree...? There's not tens of common HTTP codes for us to be just using 200, 301, 400 and 500...

File: vdr/cmd/cmd.go [hunk 8]
```diff
@@ -153,6 +156,51 @@ func resolveCmd() *cobra.Command {
...
+		Args:  cobra.ExactArgs(1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if !askYesNo("This wil delete the did document, are you sure?", cmd) {
```
Comment (reinkrul) [inline vdr/cmd/cmd.go:165]: did -> DID
wiull

File: vdr/vdr.go (hunk 7 above)
Comment (reinkrul) [reply vdr/vdr.go:179]: Comment is still incorrect. The document is updated to an empty document (except ID, ofc). Now I think of it, shouldn't we keep `@context` as well?

**Review by reinkrul — CHANGES_REQUESTED** [review]

File: docs/_static/vdr/v1.yaml [hunk 9]
```diff
@@ -77,13 +77,39 @@ paths:
...
+            text/plain:
+              schema:
+                type: string
+    delete:
+      summary: Deactivates a Nuts DID document according to the specification.
+      operationId: "deactivateDID"
+      tags:
+        - DID
+      responses:
+        "200":
```
Comment (woutslakhorst) [inline docs/_static/vdr/v1.yaml:109]: Without any return body, a 202 is more suitable.

File: docs/_static/vdr/v1.yaml [hunk 10]
```diff
@@ -77,13 +77,39 @@ paths:
...
         "200":
-          description: DID Document has been updated.
+          description: DID document has been updated.
           content:
             application/json+did-document:
               schema:
                  $ref: '#/components/schemas/DIDDocument'
         "400":
-          description: DID couldn't be updated because the input conflicts or is malformed.
+          description: DID document couldn't be updated because the did param was malformed or the document was deactivated.
```
Comment (woutslakhorst) [inline docs/_static/vdr/v1.yaml:86]: I agree with Rein that a 409 should be supported. A 409 indicates you should not repeat the request even with other parameters where a 400 suggests that you can repeat is with other parameters.

File: vdr/cmd/cmd_test.go [hunk 11]
```diff
@@ -240,6 +241,72 @@ func TestEngine_Command(t *testing.T) {
...
+	t.Run("deactivate", func(t *testing.T) {
+		t.Run("ok", func(t *testing.T) {
+			cmd := Cmd()
+			s := httptest.NewServer(http2.Handler{StatusCode: http.StatusOK})
+			os.Setenv("NUTS_ADDRESS", s.URL)
+			defer os.Unsetenv("NUTS_ADDRESS")
+			core.NewClientConfig().Load(cmd.Flags())
+			defer s.Close()
+
+			inBuf := new(bytes.Buffer)
```
Comment (woutslakhorst) [inline vdr/cmd/cmd_test.go:254]: lots of duplicate code for setting up the buffer. Could be placed in a func?

File: vdr/types/common.go [hunk 12]
```diff
@@ -53,6 +53,8 @@ type DocumentMetadata struct {
 	TimelineID hash.SHA256Hash `json:"timelineID"`
 	// Hash of DID document bytes. Is equal to payloadHash in network layer.
 	Hash hash.SHA256Hash `json:"hash"`
+	// Deactivated indicates if the document is deactivated
+	Deactivated bool `json:"deactivated"`
```
Comment (woutslakhorst) [inline vdr/types/common.go:57]: This property is also stored which could make it deviate from the actual state. Maybe the memory store must set/unset this property which each load?

**Review by woutslakhorst — COMMENTED** [review]
Some minor things that won't hurt if they are not changed.

File: docs/_static/vdr/v1.yaml (hunk 9 above)
Comment (stevenvegt) [reply docs/_static/vdr/v1.yaml:109]: 204, no content?

File: docs/_static/vdr/v1.yaml (hunk 10 above)
Comment (stevenvegt) [reply docs/_static/vdr/v1.yaml:86]: Ok, I 200 the 409 👍

File: vdr/types/common.go (hunk 12 above)
Comment (stevenvegt) [reply vdr/types/common.go:57]: This is on purpose. The intention was to deactivate the document. If the form of a deactivated document changes in the future, this should not change the state of an already deactivated document.

File: vdr/cmd/cmd_test.go (hunk 11 above)
Comment (stevenvegt) [reply vdr/cmd/cmd_test.go]: Fixed 👍

**Review by reinkrul — APPROVED** [review]

**Pull request merged**